### DISCUSSION TITLE
Uploaded Cropper image not deleted properly, but attachment file deleted just fine

### DISCUSSION THREAD

**Discussion by AndhikaWB** [orchidsoftware/platform#2320] [Q&A] [answered]
I'm creating a simple blog post using the documentation from these pages:
1. https://orchid.software/en/docs/quickstart-crud/
2. https://orchid.software/en/docs/quickstart-files/
3. https://orchid.software/en/docs/attachments/#remove

And here's my final form (slightly modified, using `Feature` model instead of `Post`):
![image](https://user-images.githubusercontent.com/17119394/181662170-f3c44d7a-410d-4eee-9a7f-d83438c127ca.png)

When I uploaded a feature icon (`Cropper`) and an attachment file (`Upload`) then created the post, all records are saved on `attachment`, `attachmenttables` (database), and `storage` folder successfully.
![image](https://user-images.githubusercontent.com/17119394/181664756-73f87cc6-d17f-4b88-8c63-0d1eaca68a96.png)
![image](https://user-images.githubusercontent.com/17119394/181664910-b1ace6d8-7699-4ec9-bcb4-e23ef0361347.png)
![image](https://user-images.githubusercontent.com/17119394/181664982-b046ae28-7bd1-4931-aa44-70769906d97b.png)
![image](https://user-images.githubusercontent.com/17119394/181665112-476468c0-40c3-42eb-ae2c-554b05751b35.png)

But when I deleted the post, only the attachment file (`Upload`) on `storage` folder is deleted, the one uploaded using `Cropper` (feature icon) is not deleted (although all the records on the database are deleted successfully).
![image](https://user-images.githubusercontent.com/17119394/181665510-e17c7ddf-9b95-4184-9698-9259a6911db9.png)

The `PostEditScreen.php` codes are pretty much like this:

```
// Form elements
Input::make('feature.title')
    ->title('Title')
    ->placeholder('Short name representing the feature')
    ->required(),

Cropper::make('feature.icon')
    ->title('Feature Icon')
    ->targetId()
    ->required(),

TextArea::make('feature.description')
    ->title('Description')
    ->rows(3)
    ->maxlength(200)
    ->placeholder('A little more explanation about the feature')
    ->required(),

Relation::make('feature.author')
    ->title('Author')
    ->fromModel(User::class, 'name')
    ->required(),

Upload::make('feature.attachment')
    ->title('Miscellaneous Files')

// Delete function
$feature->getIcon()->delete();
$feature->attachment->each->delete();
```

And the `Feature.php` model codes are pretty much like this:
```
protected $fillable = [
        'title',
        'icon',
        'description',
        'author'
];

public function getIcon()
{
        return $this->hasOne(Attachment::class, 'id', 'icon')->withDefault();
}
```

My question:
1. How do I make it to delete the `Cropper` image properly too? The example from documentation page doesn't seem to do this.
2. And if I uploaded a temporary image/file and discarded the post (not completing the edit), the image/file also retained on `storage` and `attachments` table. Is there a simple way to workaround this? I'm thinking about getting the temporary uploaded file path but I don't know how to do it cleanly without using raw query (DB::)

**Comment by AndhikaWB** [comment]
Nevermind, I can delete uploaded `Cropper` image by adding this function on the Feature model:

```
    // Get icon from attachments table
    public function getIcon()
    {
        return $this->hasOne(Attachment::class, 'id', 'icon')->withDefault();
    }

    // Get icon relative URL (/storage/xx/xx/xx/yyyyyyy.xyz)
    public function getIconRelativeUrl()
    {
        $icon = $this->getIcon()->first();
        return "/storage/{$icon->path}{$icon->name}.{$icon->extension}";
    }
```

And adding this code on delete function (FeatureEditScreen):
```
    public function remove(Feature $feature)
    {
        // Delete icon file and attachments record
        unlink(public_path() . $feature->getIconRelativeUrl());
        $feature->getIcon()->delete();

        // Delete attachmenttable record
        $feature->attachment->each->delete();

        // Delete the model itself
        $feature->delete();

        Alert::info('You have successfully deleted a feature.');

        return redirect()->route('platform.feature.list');
    }
```

As for the second question I haven't found an simple way yet.

**Comment by AndhikaWB** [comment, marked as answer]
After testing it for few days, I realized that the image and attachment system will calculate file hash to check if a file already exist on the server or not. In my case, it seems that the file (from a removed post) is still used by another post on my blog so it wasn't deleted automatically. Using above solution by me will delete the image, making all other posts who require the same image broken.

TLDR: Image and attachment will only be deleted if no other components use it.

Marking this question as answered.
